cpp/src/parquet/column_reader_test.cc

```diff
@@ -572,5 +576,615 @@ TEST_F(TestPrimitiveReader, TestNonDictionaryEncodedPagesWithExposeEncoding) {
+    ASSERT_THAT(read_reps, ElementsAre(0, 0));
+  }
+}
+
+// Test that Skip works on ByteArrays. Specifically, this is testing
+// ReadAndThrowAwayValues for ByteArrays.
+TEST(RecordReaderTest, SkipByteArray) {
+  internal::LevelInfo level_info;
+  level_info.def_level = 1;
+  level_info.rep_level = 0;
+
+  // Must use REPEATED to excercise ReadAndThrowAwayValues for ByteArrays. It
+  // does not do any buffering for Optional or Required fields as it calls
+  // ResetValues after every read.
+  NodePtr type = schema::ByteArray("b", Repetition::OPTIONAL);
+  const ColumnDescriptor descr(type, level_info.def_level, level_info.rep_level);
+
+  std::vector<std::shared_ptr<Page>> pages;
+  std::unique_ptr<PageReader> pager;
+
+  int levels_per_page = 90;
+  int num_pages = 1;
+
+  std::vector<int16_t> def_levels;
+  std::vector<int16_t> rep_levels;
+  std::vector<ByteArray> values;
+  std::vector<uint8_t> buffer;
+
+  MakePages<ByteArrayType>(&descr, num_pages, levels_per_page, def_levels, rep_levels,
+                           values, buffer, pages, Encoding::PLAIN);
+
+  pager.reset(new test::MockPageReader(pages));
+
+  std::shared_ptr<internal::RecordReader> record_reader =
+      internal::RecordReader::Make(&descr, level_info);
+  record_reader->SetPageReader(std::move(pager));
+
+  // Read one-third of the page.
+  ASSERT_EQ(record_reader->ReadRecords(/*num_records=*/30), 30);
+
+  // Skip 30 records.
+  ASSERT_EQ(record_reader->SkipRecords(/*num_records=*/30), 30);
+
+  // Read 60 more records. Only 30 will be read, since we read 30 and skipped 30,
+  // so only 30 is left.
+  ASSERT_EQ(record_reader->ReadRecords(/*num_records=*/60), 30);
+
+  auto binary_reader = dynamic_cast<BinaryRecordReader*>(record_reader.get());
+  ASSERT_NE(binary_reader, nullptr);
+  // Chunks are reset after this call.
+  ::arrow::ArrayVector array_vector = binary_reader->GetBuilderChunks();
+  ASSERT_EQ(array_vector.size(), 1);
+  auto binary_array = dynamic_cast<::arrow::BinaryArray*>(array_vector[0].get());
+  ASSERT_NE(binary_array, nullptr);
+  ASSERT_EQ(binary_array->length(), 60);
+
+  // Our values above are not spaced, however, the RecordReader will
+  // read spaced for nullable values.
+  // Create spaced expected values.
+  std::vector<std::string_view> expected_values;
+  size_t values_index = 0;
+  for (int i = 0; i < 90; ++i) {
+    if (def_levels[i] == 0) {
+      expected_values.emplace_back();
+      continue;
+    }
+    expected_values.emplace_back(reinterpret_cast<const char*>(values[values_index].ptr),
+                                 values[values_index].len);
+    ++values_index;
+  }
+
+  // Check that the expected values match the actual values.
+  for (size_t i = 0; i < 30; ++i) {
+    ASSERT_EQ(expected_values[i].compare(binary_array->GetView(i)), 0);
+  }
+  // Repeat for the next range that we read.
+  for (size_t i = 60; i < 90; ++i) {
+    ASSERT_EQ(expected_values[i].compare(binary_array->GetView(i - 30)), 0);
+  }
```

Comment:
Shouldn't we also test the validity bits of the binary array?

Reply:
Done.